crates/hulk_nao/src/microphones.rs

```diff
@@ -17,32 +21,28 @@ pub struct Microphones {
 
 impl Microphones {
     pub fn new(parameters: Parameters) -> Result<Self> {
-        let device = PCM::new("default", Direction::Capture, false)
-            .wrap_err("failed to open audio device")?;
-        {
-            let hardware_parameters =
-                HwParams::any(&device).wrap_err("failed to create hardware parameters")?;
-            hardware_parameters
-                .set_access(parameters.access)
-                .wrap_err("failed to set access")?;
-            hardware_parameters
-                .set_format(parameters.format)
-                .wrap_err("failed to set format")?;
-            hardware_parameters
-                .set_rate_near(parameters.sample_rate, ValueOr::Nearest)
-                .wrap_err("failed to set sample rate")?;
-            hardware_parameters
-                .set_channels(parameters.number_of_channels as u32)
-                .wrap_err("failed to set channel")?;
-            device
-                .hw_params(&hardware_parameters)
-                .wrap_err("failed to set hardware parameters")?;
-        }
-        device.prepare().wrap_err("failed to prepare device")?;
+        let device = create_device(&parameters).wrap_err("failed to create device")?;
         Ok(Self { device, parameters })
     }
 
-    pub fn read_from_microphones(&self) -> Result<Samples> {
+    pub fn retrying_read(&mut self) -> Result<Samples> {
+        let number_of_retries = 5;
+        for _ in 0..number_of_retries {
+            match self.read() {
+                Ok(samples) => return Ok(samples),
+                Err(error) => {
+                    warn!("failed to read from microphones: {error:#?}");
+                    self.device =
+                        create_device(&self.parameters).wrap_err("failed to create device")?;
+                }
+            }
+        }
```

Comment:
Might be fine, not sure, but if this fails repeatedly in the future, maybe add a sleep in the error branch here before creating the new device.

Also, you `?` return the error if device creation fails, I think its unlikely that after a disconnect, reading fails but opening the device anew does not.
You probably want to `continue` instead if device creation fails.

Reply:
Yes, good idea, added a sleep duration and only warn on fail to open device